searx/result_types/code.py

```diff
@@ -0,0 +1,190 @@
+        """
+        return hash(f"{self.url} {self.codelines}")
+
+    def get_lexer(self):
+        if self.code_language != "<guess>":
+            return get_lexer_by_name(self.code_language)
+
+        lexer = None
+        src_code = "\n".join([l[1] for l in self.codelines])
+
+        if self.filename:
+            try:
+                lexer = guess_lexer_for_filename(self.filename, src_code)
+            except ClassNotFound:
+                lexer = None
+
+        if not lexer:
+            try:
+                lexer = guess_lexer(src_code)
+            except ClassNotFound:
+                lexer = get_lexer_by_name("text")
+        return lexer
+
+    def HTML(self, **options) -> str:  # pyright: ignore[reportUnknownParameterType, reportMissingParameterType]
+        """Rendered HTML, additional options are accepted, for more details have
+        a look at HtmlFormatter_.
+
+        .. _HtmlFormatter: https://pygments.org/docs/formatters/#HtmlFormatter
+        """
+        lexer = self.get_lexer()
+
+        line_no: int = 0  # current line number
+        last_line_no: int | None = None
+        code_block_start: int = 0  # line where the current code block starts
+        code_block: list[str] = []  # lines of the current code block
+        html_code_blocks: list[str] = []  # HTML representation of all code blocks
+
+        def _render(**kwargs):  # pyright: ignore[reportUnknownParameterType, reportMissingParameterType]
+            for k, default in [
+                ("linenos", "inline"),
+                ("linenostart", code_block_start),
+                ("cssclass", "code-highlight"),
+                ("hl_lines", [hl - code_block_start + 1 for hl in self.hl_lines]),
+            ]:
+                kwargs[k] = kwargs.get(k, default)  # pyright: ignore[reportUnknownMemberType]
+
+            # Wrap the code inside <pre> blocks using <code>, as recommended by
+            # the HTML5 specification (default is False).  Do we need this?
+            #
+            # kwargs["wrapcode"] = kwargs.get("wrapcode", True)
+
+            html_code_blocks.append(
+                highlight(
+                    "\n".join(code_block),
+                    lexer,
+                    HtmlFormatter(**kwargs),  # pyright: ignore[reportUnknownArgumentType]
+                )
+            )
+
+        for line_no, code_line in self.codelines:
+            if last_line_no is None:
+                # initial start condition
+                code_block_start = line_no
+
+            if last_line_no is not None and last_line_no + 1 != line_no:
+                # new code block is detected, render current code block
+                _render(**options)  # pyright: ignore[reportUnknownArgumentType]
+                # reset conditions for next code block, which first line is the
+                # current code line
+                code_block = [code_line]
+                code_block_start = line_no
+                last_line_no = line_no
```

Comment:
```suggestion
                previous_line_no = line_no
```
I've struggled a bit with understanding this code initially because I thought that `last_line_no` is meant to represent the last number of all codelines, e.g. (`6` if we have codelines `1,2,...,6`.

Reply:
Yes, the name is miss leading .. this is a pointer to the end of the *current* code block, I renamed the variable to: 

```
 code_block_end: int | None = None  # line where the current code ends
```